### PR TITLE
refactor: reorganize global flags

### PR DESCRIPTION
**Pull request by ry**
- Remove ability to specify run arguments like `--allow-net` after the script argument. It's too hacky to make work with clap.
- Prevent clap from auto-detecting terminal width and attempting to wrap text.
- Remove `--v8-options`, instead use `--v8-flags=--help`
- Give more descriptive names to unit tests in flags.rs
- Assume argv and subcommand into DenoFlags struct so the output of flags module is only DenoFlags rather than the tuple (subcommand, flags, argv).
- Improve CLI help text
- Make `deno run` specific args like `--allow-net` only show up in 'deno
  help run' instead of as global flags in `deno help`.
- Removes `deno version` to simplify our implementation and be closer to clap defaults. `deno -V` now only shows Deno's version and not V8's nor TypeScript. `Deno.versions` can be used to see that information.

### CODE REVIEW

**Comment by ry** [comment]
@nayeemrmn Why would `--config` apply to fetch? `deno fetch --no-fetch $URL` doesn't sound  right. I agree about `--lock` and `--lock-write`.

**Comment by nayeemrmn** [comment]
@ry 

> Why would `--config` apply to fetch?

Doesn't it affect compilation?

> `deno fetch --no-fetch $URL` doesn't sound right.

https://github.com/denoland/deno/pull/3383#discussion_r348481787. I agree that the naming is inconsistent. That was actually what prompted #3384.

**Comment by ry** [comment]
> Doesn't it affect compilation?

Yes, ok. I agree and I've made the change.

Regarding `deno fetch --no-fetch` ... I don't understand your reasoning still.

**Comment by nayeemrmn** [comment]
It's hard to reason about since I'm not sure what the use case of `--no-fetch` is in the first place...

Say I want to disable remote module downloading but I still want to use `deno fetch` to _compile_ ahead of time. If `deno fetch` is forced to download everything, `--no-fetch` for `deno run` becomes useless since it's all already cached. I don't see why these two features should be at odds.

**Comment by ry** [comment]
@nayeemrmn Ah ok. I understand now. But maybe we should call it something else like `--recompile-only` ? In any case let's do that in a follow up PR since it should have tests. This one's already getting too big.

**Comment by nayeemrmn** [comment]
@ry None of this will matter after #3384 😅 I suggest just applying it as-is for now and certainly not introducing another flag.

File: cli/flags.rs [hunk 1]
```diff
@@ -164,192 +190,378 @@ To get help on the another subcommands (run in this case):
...
+  let mut buf: Vec<u8> = vec![];
+  use std::str::FromStr;
+  clap_root().gen_completions_to(
+    "deno",
+    clap::Shell::from_str(shell).unwrap(),
+    &mut buf,
+  );
+  // TODO(ry) This flags module should only be for parsing flags, not actually
+  // acting upon the flags. Although this print is innocent, it breaks the
+  // model. The print should be moved out.
```
Comment (bartlomieju) [inline cli/flags.rs]: :+1:

File: cli/flags.rs [hunk 2]
```diff
@@ -581,735 +810,502 @@ Demonstrates breaking the input up by space delimiter instead of by lines:
...
-
-        let (script_args, flags_) = parse_script_args(script_args, flags);
-        flags = flags_;
-        argv.extend(script_args);
-      }
-      DenoSubcommand::Run
+// TODO move this to utility module...
+// TODO what does this function even do?
+pub fn resolve_urls(urls: Vec<String>) -> Vec<String> {
+  use url::Url;
```
Comment (bartlomieju) [inline cli/flags.rs]: Yeah, this is very specific to cache blacklist - generally it strips `#fragment` part of URL. I'm not sure though why it pops last slash /shrug

File: cli/flags.rs [hunk 3]
```diff
@@ -581,735 +810,502 @@ Demonstrates breaking the input up by space delimiter instead of by lines:
...
-        reload: true,
-        ..DenoFlags::default()
-      }
-    );
-    assert_eq!(subcommand, DenoSubcommand::Run);
-    assert_eq!(argv, svec!["deno", "script.ts"]);
+  out
+}
+
+fn arg_hacks(mut args: Vec<String>) -> Vec<String> {
```
Comment (bartlomieju) [inline cli/flags.rs]: Nice

File: cli/flags.rs [hunk 4]
```diff
@@ -56,106 +74,114 @@ pub struct DenoFlags {
   // Use tokio::runtime::current_thread
   pub current_thread: bool,
 
+  pub bundle_output: Option<String>,
```
Comment (bartlomieju) [inline cli/flags.rs]: It's strange that this is part of `DenoFlags` - it's very specific to `deno bundle`, why not put it into `argv`?

**Comment by hayd** [comment]
If we were to rename `fetch` to `compile` I think the flag would make a lot more sense i.e. `deno compile --no-fetch`.

I know @ry you wanted `deno compile` to mean "compile into a single binary including deno itself" but perhaps that would be better as `deno binary` or something like that...

File: cli/flags.rs (hunk 4 above)
Comment (ry) [reply cli/flags.rs]: I actually don't like the fact that so many subcommands use argv - I think it should be restricted to the "Run" subcommand.  I added some TODOs elsewhere in this patch that argv should become an `Option<Vec<String>>`. The worry is that it's being used as a general purpose way to pass data out.

The case of bundle, it does run JS code but has some bits of rust code to help it, so it's not a real Run command. I removed this bit from lib.rs where argv is being used inappropriately 
```rust
  assert!(state.argv.len() >= 2);	
  let out_file = match state.argv.len() {	
    3 => Some(state.argv[2].clone()),	
    _ => None,	
  };
```

File: cli/flags.rs [hunk 5]
```diff
@@ -56,106 +74,114 @@ pub struct DenoFlags {
...
+
+  let mut flags = DenoFlags::default();
+
+  flags.argv.push(args0);
+
+  if matches.is_present("log-level") {
+    flags.log_level = match matches.value_of("log-level").unwrap() {
+      "debug" => Some(Level::Debug),
+      "info" => Some(Level::Info),
+      _ => unreachable!(),
```
Comment (piscisaureus) [inline cli/flags.rs]: Something better than "unreachable!" here? `deno --log-level warn` now begets this nasty crash.

Comment (piscisaureus) [reply cli/flags.rs]: nvm, doesn't actually seem to happen

File: cli/flags.rs (hunk 2 above)
Comment (ry) [reply cli/flags.rs]: Thanks - added a little doc line.

**Comment by ry** [comment]
> That said it's a very long patch. Are there areas in particular that you think need some extra eyeballs?

Nope - I think this is good to go

> Also, has this been addressed? #3389 (comment)

Now it has.

**Pull request merged**